jwe/internal/aescbc/aescbc.go

```
@@ -125,7 +127,12 @@ func (c Hmac) Seal(dst, nonce, plaintext, data []byte) []byte {
 	cbc := cipher.NewCBCEncrypter(c.blockCipher, nonce)
 	cbc.CryptBlocks(ciphertext, ciphertext)
 
-	authtag := c.ComputeAuthTag(data, nonce, ciphertext)
+	authtag, err := c.ComputeAuthTag(data, nonce, ciphertext)
+	if err != nil {
+		// Hmac implement cipher.AEAD interface. Seal can't return error
+		// TODO: How to fix it? print log?
+		return nil
+	}
```

I think that Write method doesn't fail in ComputeAuthTag.
Is it ok just print log?

---

Well, if you changed the signature we should propagate the error. If ComputeAuthTag indeed never returns an error, then we shouldn't be adding the second return value, no?

---

I didn't explain it well enough.

Hmac implement standard library cipher.AEAD interface. I can't change it.

https://golang.org/src/crypto/cipher/gcm.go#L33

---

Yeah, but previously  `ComputeAuthTag` didn't have an error returned, right? You added it. If you added it, you have to handle it. If you think that it doesn't need to be handled, then you should change `ComputeaAuthTag` back to the error-less version.

---

Sorry in Japanse.

1. ComputeAuthTagメソッドの中のWriteのエラーハンドリングをするべきかどうか

ComputeAuthTagメソッドの中のWriteはerrorが 返ってくる可能性があるので、エラーハンドリングはするべきだと思います。
現在の実装ではerrorは返ってきませんが、将来的に返ってこない保証はないためです。

https://github.com/golang/go/blob/bce1e25b717e07e05174117af6907999b3f51285/src/crypto/sha256/sha256.go#L184

https://github.com/golang/go/blob/bce1e25b717e07e05174117af6907999b3f51285/src/crypto/sha512/sha512.go#L255


2. ComputeAuthTagはerrorを返すべきか

Hmac構造体のComputeAuthTagの利用者がerrorのハンドリングの選択権を得るためにも返した方がいいと思います。
実際Hmac構造体のOpenメソッドがComputeAuthTagメソッドを実行していますが、こちらではerrorを返すことが可能なのでComputeAuthTagから返ってきたエラーをハンドリンスした方がいいと思いました。
https://github.com/sonatard/jwx/blob/ec524a3a4592dfd6b08bf059961c84c744bfc948/jwe/internal/aescbc/aescbc.go#L169

(こちらはエラーハンドリングの実装を忘れていて、先ほどコミットしました。  https://github.com/lestrrat-go/jwx/pull/177/commits/ec524a3a4592dfd6b08bf059961c84c744bfc948

Hmac構造体のSealメソッドの都合でエラーを返さないのであれば、Sealメソッドの中で明示的にerrorを返さない理由を説明するとともに対処するのがいいと思いました。

とはいえ、現実的にerror返ってこないのだからWriteでerrorを受けないことをコメントで明示するだけでもいいと思います。

ご意見待ってます！

---

基本的にはエラーは返せるギリギリまで返し続けるほうがいいと思うので、自分たちで実装しているensureSizeはエラーを返すべきだと思います。なので、このコメントが書かれたensureSize内のコード部分ではエラーを処理（返す）べきだと思っています。

Hmac.Sealでエラーを無視するならわかります。これはインターフェースが許さないから、という明らかな理由があるので説明を書いてエラーを無視、は問題ないかとおもいます。

ちなみにensureSizeが失敗するとそのあと現実的にはそのあと処理が進まないのだから、こここではひょっとしてpanicすべきなのかも…？

---

s/ここ/Seal/

---

ありがとうございます！
error状態で後続に処理の続けてしまうと危険そうなのでpanicが良さそうですね！

現在コメントが書かれているメソッドはHmac.Sealメソッドなので、ここにコメントを書いてpanicにしようと思います！

https://github.com/sonatard/jwx/blob/ec524a3a4592dfd6b08bf059961c84c744bfc948/jwe/internal/aescbc/aescbc.go#L121-L135